<div style="display: flex; align-items: center; gap: 20px; font-family: Arial, sans-serif;">
    <div style="margin-right: 30px;">
        <img src="../Sources/Logo.png" width="300" style="border-radius: 10px;"/>
    </div>
    <div style="line-height: 1.5;">
        <strong style="font-size: 18px; display: block; margin-bottom: 5px;">
            <a href='https://t.me/M4m4ev_Alex' style="text-decoration: none; color: #1a73e8;">
                Репетитор по информатике | Мамаев Алексей
            </a>
        </strong>
        <span style="font-size: 16px; color: #333;">
            Разбор <a href='https://euler.jakumo.org/problem/5-naimensee-kratnoe' style="text-decoration: none; color: #1a73e8;">
            5-й задачи</a> проекта Эйлера
        </span>
        <span style="font-size: 16px; color: #333;">
            <br><br><br><br><br><br>ТГК: <a href='https://t.me/M4m4ev_Alex' style="text-decoration: none; color: #1a73e8;">@M4m4ev_Alex</a><br>
        </span>
    </div>
</div>

### Задача 5

Число `2520` является наименьшим положительным числом, которое делится без остатка на все числа от `1` до `10`.

Необходимо найти **наименьшее положительное число**, которое делится нацело на все числа от `1` до `20`.

Ответ: 232792560

## Используемые библиотеки

В данном разборе используются стандартные библиотеки Python.

- `perf_counter` — измерение времени выполнения алгоритмов.
- `lcm` — вычисление наименьшего общего кратного.
- `prod` — вычисление произведения элементов последовательности.
- `reduce` — последовательное применение функции ко всем элементам коллекции.

In [12]:
from time import perf_counter
from math import lcm, prod
from functools import reduce

## 1) Через разложение на простые множители

В данном решении задача сводится к поиску **наименьшего общего кратного (НОК)** всех чисел от `1` до `n`.

Идея заключается в том, чтобы разложить каждое число диапазона на простые множители и для каждого простого множителя сохранить **максимальную степень**, с которой он встречается. Именно произведение этих множителей в максимальных степенях образует наименьшее число, кратное всем числам диапазона.

### Алгоритм

1. Функция `get_divs()` раскладывает число на простые множители.
2. Для каждого числа от `1` до `n` выполняется его разложение.
3. В словаре `d` для каждого простого множителя сохраняется максимальная степень.
4. После обработки всех чисел вычисляется произведение всех простых множителей в найденных степенях.
5. Полученное значение возвращается в качестве ответа.

### Особенности

- Основан на разложении чисел на простые множители.
- Использует словарь для хранения максимальных степеней простых множителей.
- Функция `prod()` позволяет компактно вычислить итоговое произведение.
- Легко обобщается для любого значения `n`.
- Не требует перебора возможных ответов.

### Вывод

Использование разложения на простые множители позволяет напрямую построить наименьшее общее кратное всех чисел диапазона. Такой подход сочетает математическую строгость, наглядность и достаточно высокую эффективность.

In [17]:
def get_divs(n: int) -> dict[int, int]:
    r = {}
    div = 2
    while n > 1:
        if n % div == 0:
            r[div] = r.get(div, 0) + 1
            n //= div
        else:
            div += 1
    return r
        
def get_result_1(n: int) -> int:
    d = {}
    for i in range(1, n + 1):
        t = get_divs(i)
        for j in t:
            d[j] = max(d.get(j, 0), t[j])
            
    return prod(i**d[i] for i in d)


start = perf_counter()
print(get_result_1(20))
end = perf_counter()

print(f'{end - start:.5f}')

232792560
0.00230


## 2) Последовательный поиск наименьшего общего кратного

В отличие от предыдущих решений, здесь не используется разложение чисел на простые множители.

Алгоритм начинает работу с максимального числа диапазона и последовательно строит **наименьшее общее кратное** для всех чисел от `20` до `1`.

На каждой итерации текущее значение увеличивается на свою первоначальную величину до тех пор, пока не станет кратным очередному числу.

### Алгоритм

1. Начальное значение `s` принимается равным `mx`.
2. Последовательно перебираются числа от `mx` до `1`.
3. Для каждого числа сохраняется текущее значение `s` в переменную `tmp`.
4. Пока `s` не делится на текущее число без остатка, к нему прибавляется `tmp`.
5. После обработки всех чисел значение `s` является искомым наименьшим общим кратным.

### Особенности

- Не требует разложения чисел на простые множители.
- Использует последовательное построение наименьшего общего кратного.
- Реализация достаточно компактна и легко читается.
- На каждой итерации выполняются только операции сложения и проверки делимости.
- Подходит для небольших значений `n`, однако при увеличении диапазона работает значительно медленнее математического метода.

### Вывод

Данный подход позволяет найти ответ без использования дополнительных математических преобразований. Несмотря на простоту реализации, по эффективности он уступает решению через разложение на простые множители, так как требует значительно большего количества операций.

In [22]:
def get_result_2(mx: int) -> int:
    s = mx
    
    for i in range(mx, 0, -1):
        tmp = s
        while s % i != 0:
            s += tmp
    return s
    
    
start = perf_counter()
print(get_result_2(20))
end = perf_counter()

print(f'{end - start:.5f}')

232792560
0.00041


## 3) Через НОД и НОК

В этом подходе задача решается через последовательное вычисление **наименьшего общего кратного** для чисел от `1` до `20`.

Используется важная связь между НОД и НОК двух чисел:

```text
НОК(a, b) = a · b / НОД(a, b)
```

Сначала вычисляется НОК для первых двух чисел, затем результат объединяется со следующим числом, потом со следующим и так далее.

### Алгоритм

1. Функция `get_NOD(a, b)` вычисляет наибольший общий делитель двух чисел с помощью алгоритма Евклида.
2. Функция `get_NOK(a, b)` вычисляет наименьшее общее кратное двух чисел через НОД.
3. Функция `reduce()` последовательно применяет `get_NOK` ко всем числам от `1` до `n`.
4. В результате получается число, которое делится на все числа диапазона.

### Особенности

- Решение напрямую использует связь между НОД и НОК.
- Алгоритм Евклида позволяет эффективно находить НОД.
- `reduce()` удобно применяет операцию НОК ко всему диапазону чисел.
- Код получается компактным и хорошо масштабируется для других значений `n`.
- Это один из самых аккуратных и математически естественных способов решения задачи.

### Вывод

Использование НОД и НОК позволяет получить короткое и эффективное решение. Такой подход хорошо подходит для задач, где необходимо найти число, кратное сразу нескольким значениям.

In [19]:
def get_NOD(a: int, b: int) -> int:
    while a > 0 and b > 0:
        if a > b:
            a %= b
        else:
            b %= a
    return max(a, b)

def get_NOK(a: int, b: int) -> int:
    return (a * b) // get_NOD(a, b)
    
def get_result_3(n: int) -> int:
    return reduce(get_NOK, range(1, n + 1))

start = perf_counter()
print(get_result_3(20))
end = perf_counter()

print(f'{end - start:.5f}')

232792560
0.00033


## 4) Использование встроенной функции `lcm()`

В данном решении задача полностью передаётся встроенной функции `lcm()` из модуля `math`.

Функция принимает сразу весь диапазон чисел от `1` до `n` и вычисляет их наименьшее общее кратное.

Для передачи последовательности в качестве отдельных аргументов используется оператор распаковки `*`.

### Алгоритм

1. Формируется диапазон чисел от `1` до `n`.
2. С помощью оператора `*` элементы диапазона передаются функции `lcm()` как отдельные аргументы.
3. Функция `lcm()` вычисляет наименьшее общее кратное всех переданных чисел.
4. Полученное значение возвращается в качестве ответа.

### Особенности

- Используется встроенная функция `lcm()` из стандартной библиотеки Python.
- Применяется оператор распаковки `*` для передачи диапазона чисел.
- Код получается максимально компактным и легко читается.
- Не требуется самостоятельно реализовывать алгоритмы вычисления НОД или НОК.
- Встроенная реализация оптимизирована и обычно работает быстрее аналогичной реализации на чистом Python.

### Вывод

Использование функции `lcm()` совместно с оператором распаковки позволяет получить наиболее короткое и понятное решение задачи. Благодаря возможностям стандартной библиотеки Python весь алгоритм сводится к одной строке кода без потери эффективности.

In [20]:
def get_result_4(n: int) -> int:
    return lcm(*range(1, n + 1))

start = perf_counter()
print(get_result_4(20))
end = perf_counter()

print(f'{end - start:.5f}')

232792560
0.00012


## Сравнение времени выполнения

Для сравнения производительности используется функция `timeit()`, предназначенная специально для измерения времени выполнения небольших участков кода.

Каждый алгоритм запускается **1000 раз**, после чего вычисляется **среднее время выполнения одного запуска**.

Измерения проводятся для трёх различных значений параметра `n` (`20`, `100` и `1000`). Это позволяет сравнить не только скорость работы алгоритмов, но и оценить, как изменяется их производительность при увеличении размера входных данных.

In [23]:
from timeit import timeit

funcs = (
    get_result_1,
    get_result_2,
    get_result_3,
    get_result_4,
)

for j in (20, 100, 1000):
    print(f'n = {j}')
    for i, func in enumerate(funcs, start=1):
        tm = timeit(lambda: func(j), number=1000) / 1000
        print(f'Алгоритм {i}: {tm:.7f}')
    print()

n = 20
Алгоритм 1: 0.0000208
Алгоритм 2: 0.0000049
Алгоритм 3: 0.0000068
Алгоритм 4: 0.0000014

n = 100
Алгоритм 1: 0.0001952
Алгоритм 2: 0.0000811
Алгоритм 3: 0.0000420
Алгоритм 4: 0.0000130

n = 1000
Алгоритм 1: 0.0097944
Алгоритм 2: 0.0213513
Алгоритм 3: 0.0007279
Алгоритм 4: 0.0004048



# Заключение

В ходе решения задачи были рассмотрены четыре различных подхода к поиску наименьшего числа, которое делится нацело на все числа заданного диапазона.

Первый алгоритм основан на разложении чисел на простые множители. Он позволяет напрямую построить наименьшее общее кратное, используя максимальные степени простых множителей. Такой подход хорошо демонстрирует математическую идею решения задачи.

Во втором решении наименьшее общее кратное строится последовательным увеличением текущего значения до тех пор, пока оно не станет кратно очередному числу диапазона. Несмотря на простоту реализации, этот метод плохо масштабируется и при увеличении `n` начинает заметно уступать другим решениям.

Третий алгоритм использует вычисление наименьшего общего кратного через наибольший общий делитель. Благодаря алгоритму Евклида он показывает хорошую производительность даже при больших значениях `n`.

Наиболее эффективным оказалось четвёртое решение, использующее встроенную функцию `lcm()` из стандартной библиотеки Python. Оно оказалось самым быстрым на всех рассмотренных диапазонах, а его код является наиболее компактным.

Среднее время выполнения алгоритмов:

| `n` | Алгоритм 1 | Алгоритм 2 | Алгоритм 3 | Алгоритм 4 |
|---:|---:|---:|---:|---:|
| **20** | `0.0000208` с | `0.0000049` с | `0.0000068` с | `0.0000014` с |
| **100** | `0.0001952` с | `0.0000811` с | `0.0000420` с | `0.0000130` с |
| **1000** | `0.0097944` с | `0.0213513` с | `0.0007279` с | `0.0004048` с |

Результаты показывают, что для небольших значений `n` различия между алгоритмами незначительны, и второй подход даже немного опережает третье решение. Однако с увеличением размера диапазона его производительность резко снижается.

Третий и четвёртый алгоритмы значительно лучше масштабируются, а встроенная функция `lcm()` демонстрирует наилучшие результаты на всех проведённых тестах. Это ещё раз показывает, что при выборе алгоритма важно учитывать не только его скорость на небольших примерах, но и поведение при увеличении объёма входных данных.